### PR TITLE
DEPR: Int64Index, UInt64Index & Float64Index

### PR DESCRIPTION
**Pull request by topper-123**
* [x] xref #42717
* [x] xref #41153 & #42706

Deprecates Int64Index, UInt64Index & Float64Index.

I don't think it will be sensible to emit a deprecation warning for these because they have a lot of internal usage as is, so I've just added texts explaining that these are deprecated and will be removed in Pandas 2.0.

EDIT: it does now emit an deprecation warning when accessed from the main namespace. See changes to `pandas.__init__.py` for implementation.

```python
>>> pd.Int64Index
FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.NumericIndex with the appropriate dtype instead.
```

### CODE REVIEW

File: pandas/core/indexes/numeric.py [hunk 1]
```diff
@@ -42,51 +42,44 @@
...
-    of `Index` with purely %(ltype)s labels. %(extra)s.
+    storing axis labels for all pandas objects. NumericIndex is a special case
+    of `Index` with purely numpy int/uint/float labels.
+
+    .. versionadded:: 1.4.0
+
+    Notes
+    -----
+    An NumericIndex instance can **only** contain numpy int64/32/16/8, uint64/32/16/8 or
+    float64/32/16 dtype. In Particulat, ``NumericIndex`` *can not* hold Pandas numeric
```
Comment (mroeschke) [inline pandas/core/indexes/numeric.py:57]: `Particulat` -> `particular`

**Comment by jbrockmendel** [comment]
xref #43002

File: pandas/core/indexes/numeric.py (hunk 1 above)
Comment (topper-123) [reply pandas/core/indexes/numeric.py]: Changed.

**Comment by lithomas1** [comment]
IMHO, we probably should emit some sort of warning to users if 2.0 is the next release. I'm a little worried since 2.0 should be the next release, and this is a somewhat big change. We can always suppress the deprecation warnings internally.

**Comment by topper-123** [comment]
I've made a new version where accessing ``Int64Index`` etc. in the main namespace emits a FutureWarning. So they give warnings now:

```python
>>> import pandas as pd
>>> idx = pd.Int64Index([8, 9])
FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.NumericIndex with the appropriate dtype instead.
>>> isinstance(idx, pd.Int64Index)
FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.NumericIndex with the appropriate dtype instead.
True
```

but this does not:

```python
>>> idx = pd.Index([8, 9])
>>> idx
Int64Index([8, 9], dtype='int64')
>>> isinstance(idx, pd.NumericIndex)
True
```

This last examples don't give warnings, because they're not deprecated, but will change meaning in Pandas 2.0 (will return a NumericIndex rather than a Int64Index and Int64Index is a subclass of Int64Index).

There warnings should  give users a good hint on where they need to update their code to stay compatible in Pandas 2.0.

**Comment by topper-123** [comment]
Got some failures still, but will work on them. The idea is completed, so commentd welcome.

File: doc/source/user_guide/advanced.rst [hunk 2]
```diff
@@ -851,11 +851,11 @@ values **not** in the categories, similarly to how you can reindex **any** panda
 Int64Index and RangeIndex
 ~~~~~~~~~~~~~~~~~~~~~~~~~
 
-.. note::
-
+.. deprecated:: 1.4.0
     In pandas 2.0, :class:`NumericIndex` will become the default index type for numeric types
     instead of ``Int64Index``, ``Float64Index`` and ``UInt64Index`` and those index types
-    will be removed. See :ref:`here <advanced.numericindex>` for more.
+    are therefore deprecated and will be removed in Pandas 2.0.
```
Comment (jreback) [inline doc/source/user_guide/advanced.rst:857]: instead of saying Pandas 2.0, just a future version (we don't explicitly list the removal versions anywhere else i believe)

Comment (jreback) [reply doc/source/user_guide/advanced.rst:857]: same comment applies to all docs below

File: pandas/__init__.py [hunk 3]
```diff
@@ -186,10 +183,32 @@
 
 
 # GH 27101
+__deprecated_num_index_names = ["Float64Index", "Int64Index", "UInt64Index"]
+
+
+def __dir__():
```
Comment (jreback) [inline pandas/__init__.py:189]: why is this needed?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/__init__.py (hunk 3 above)
Comment (topper-123) [reply pandas/__init__.py:189]: This is so Int64Index etc. show up in dir calls. Int64Index etc. have been quite prominent parts of the public API, so I think it's better that they still show up for the duration of Pandas 1.x. This should be removed in Pandas 2.0.

File: doc/source/user_guide/advanced.rst (hunk 2 above)
Comment (topper-123) [reply doc/source/user_guide/advanced.rst:857]: Ok, done.

File: pandas/__init__.py (hunk 3 above)
Comment (jreback) [reply pandas/__init__.py:189]: ok can you add a comment to this effect

File: doc/source/user_guide/advanced.rst (hunk 2 above)
Comment (jreback) [reply doc/source/user_guide/advanced.rst:857]: did you push the update?

**Comment by topper-123** [comment]
Updated.

**Comment by topper-123** [comment]
Gentle ping…

File: doc/source/user_guide/advanced.rst [hunk 4]
```diff
@@ -869,12 +869,12 @@ implementing an ordered, sliceable set.
 Float64Index
 ~~~~~~~~~~~~
 
-.. note::
-
-    In pandas 2.0, :class:`NumericIndex` will become the default index type for numeric types
+.. deprecated:: 1.4.0
```
Comment (jreback) [inline doc/source/user_guide/advanced.rst:872]: does this need a blank line after?

File: doc/source/user_guide/advanced.rst [hunk 5]
```diff
@@ -981,9 +981,9 @@ NumericIndex
 .. note::
 
     In pandas 2.0, :class:`NumericIndex` will become the default index type for numeric types
-    instead of ``Int64Index``, ``Float64Index`` and ``UInt64Index`` and those index types
-    will be removed.
-    ``RangeIndex`` however, will not be removed, as it represents an optimized version of an integer index.
+    instead of :class:`Int64Index`, :class:`Float64Index` and :class:`UInt64Index` and those index types
+    are therefore deprecated and will be removed in a futureversion.
```
Comment (jreback) [inline doc/source/user_guide/advanced.rst:985]: future version

**Review by jreback — CHANGES_REQUESTED** [review]
very small comment, pls rebase

File: doc/source/user_guide/advanced.rst (hunk 4 above)
Comment (topper-123) [reply doc/source/user_guide/advanced.rst:872]: No. It's used in in used e.g. in user_guide/io.rst, where it works fine.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @topper-123 very  nice!

**Comment by jbrockmendel** [comment]
@topper-123 want to enforce this deprecation and rip out some code?
